### DISCUSSION TITLE
mqtt.publish @berry

### DISCUSSION THREAD

**Discussion by MubiTec** [arendst/Tasmota#16157] [Q&A] [answered]
Hi,

I have a Problem. I want to publish my data to Mqtt FROM a berry Script.
When I put these lines in berry console IT works fine:
import mqtt
mqtt.publish("stat/DVES_1A83D0/result","123")

If I but them in my working autoexec Script nothing happens, no mqtt message no error. All other calculations of the function works fine, so I am Sure that the function is running.
I use tasmota 12.0.2 FROM http://ota.tasmota.com/tasmota32/release/tasmota32.bin
Any idea?

**Comment by sfromis** [comment, marked as answer]
Using MQTT functions directly in `autoexec.be` cannot work, as this script is executed before Tasmota connects to Wifi and MQTT.

**Comment by MubiTec** [comment]
Thank you for your answer. Now I tried this:

in autoexec:
load("battery.be")

in battery.be
def w(n)
var sensors
import json
import persist
import string
import mqtt
var w
sensors = json.load(tasmota.read_sensors())
if n!= nil
persist.w=n
end
w=real(sensors['AM2301']['Temperature']) 
w=w/3600

persist.w += w
if n!= nil
persist.w=n
end
return persist.w

mqtt.publish("stat/DVES_1A83D0/result","123")
#tasmota.cmd(publish stat/%topic%/RESULT {"Button1":"L"}) 
persist.save() # save to _persist.json
end
tasmota.add_cmd('w', w)

w = w()
tasmota.add_driver(w)

def every_second()
w() 
end

But it still dies not work :-(

**Reply by sfromis** [reply]
Moving the code to another file does not make it execute later, after the time to get Wifi/MQTT connection. You could consider using something `tasmota.add_rule('mqtt.connected', ...)` for a better time to run the initialize code.

What I meant with "directly" was not included or not in `autoexec.be`, but executed at that time.

**Comment by MubiTec** [comment]
Ok, I Tried this. the Battery.be has been loaded now after mqtt and WiFi Start, but without any effect :-(
Any ohter advise?

**Reply by sfromis** [reply]
Clean up your messy code :wink: 

The code you posted earlier merely defines global functions, and adds a command plus a driver which has no driver functions at all. Even has some dead code after the return. Executes nothing. Your usage of `every_second()` looks like you want to actually have a driver, which means that you should have a class instance. Applying the usual indentation rules would be a good idea, makes such issues stand out.
